datastore/warehouse/warehouse.go

```diff
@@ -0,0 +1,278 @@
+package warehouse
+
+import (
+	"compress/gzip"
+	"context"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/blocknative/dreamboat/structs"
+	"github.com/lthibault/log"
+)
+
+var (
+	fileIdleTime   = structs.DurationPerSlot
+	filesPruneTick = fileIdleTime * 3
+	ErrClosed      = errors.New("closed")
+)
+
+type Warehouse struct {
+	logger   log.Logger
+	requests chan StoreRequest
+
+	shutdown       chan struct{}
+	runningWorkers *structs.TimeoutWaitGroup
+
+	m WarehouseMetrics
+}
+
+func NewWarehouse(logger log.Logger, bufSize int) *Warehouse {
+	wh := &Warehouse{logger: logger.WithField("service", "warehouse"), requests: make(chan StoreRequest), runningWorkers: structs.NewTimeoutWaitGroup()} // channel must be unbuffered, for not accepting requests that will not be handled
+	wh.initMetrics()
+	return wh
+}
+
+func (s *Warehouse) RunParallel(ctx context.Context, datadir string, numWorkers int) error {
+	for i := 0; i < numWorkers; i++ {
+		go s.Run(ctx, datadir, i)
+	}
+
+	return nil
+}
+
+func (s *Warehouse) Run(ctx context.Context, datadir string, id int) {
+	s.runningWorkers.Add(1)
```

Comment:
may be a nitpick but call add outside :)

Reply:
I would not do that in case someone wants to call `Run()` directly, instead of `RunParallel()`